pkg/route/controller/ingress/ingress.go

```diff
@@ -0,0 +1,591 @@
+	if t > maxResourceInterval {
+		t = maxResourceInterval
+	}
+	return t, true
+}
+
+// Controller ensures that zero or more routes exist to match any supported ingress. The
+// controller creates a controller owner reference from the route to the parent ingress,
+// allowing users to orphan their ingress. All owned routes have specific spec fields
+// managed (those attributes present on the ingress), while any other fields may be
+// modified by the user.
+//
+// Invariants:
+//
+// 1. For every ingress path rule with a non-empty backend statement, a route should
+//    exist that points to that backend.
+// 2. For every TLS hostname that has a corresponding path rule and points to a secret
+//    that exists, a route should exist with a valid TLS config from that secret.
+// 3. For every service referenced by the ingress path rule, the route should have
+//    an update to date target port based on the service.
+// 4. A route owned by an ingress that no longer satisfies the first three invariants
+//    should be deleted.
+//
+// Unsupported attributes:
+//
+// * the ingress class attribute
+// * nginx annotations
+// * the empty backend
+// * paths with empty backends
+// * creating a dynamic route spec.host
+//
+type Controller struct {
+	eventRecorder record.EventRecorder
+
+	// Allows injection for testing, controls requeues on errors
+	resourceFailureDelayFn func(requeue int) (time.Duration, bool)
+
+	client        routeclient.RoutesGetter
+	secretLister  corelisters.SecretLister
+	ingressLister extensionslisters.IngressLister
+	routeLister   routelisters.RouteLister
+	serviceLister corelisters.ServiceLister
+
+	// queue is the list of namespace keys that must be synced.
+	queue workqueue.RateLimitingInterface
+
+	// syncs are the items that must return true before the queue can be processed
+	syncs []cache.InformerSynced
+}
+
+type queueKey struct {
+	namespace string
+	name      string
+}
+
+// NewController instantiates a Controller
+func NewController(eventsClient kv1core.EventsGetter, client routeclient.RoutesGetter, ingresses extensionsinformers.IngressInformer, secrets coreinformers.SecretInformer, services coreinformers.ServiceInformer, routes routeinformers.RouteInformer) *Controller {
+	broadcaster := record.NewBroadcaster()
+	broadcaster.StartLogging(glog.Infof)
+	// TODO: remove the wrapper when every clients have moved to use the clientset.
+	broadcaster.StartRecordingToSink(&kv1core.EventSinkImpl{Interface: eventsClient.Events("")})
+	recorder := broadcaster.NewRecorder(legacyscheme.Scheme, v1.EventSource{Component: "ingress-to-route-controller"})
+
+	c := &Controller{
+		eventRecorder: recorder,
+		queue:         workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "ingress-to-route"),
+
+		client:        client,
+		ingressLister: ingresses.Lister(),
+		secretLister:  secrets.Lister(),
+		routeLister:   routes.Lister(),
+		serviceLister: services.Lister(),
```

Comment:
normally, you have a sync functions collected from these for a "wait for sync before run" check.